### PR TITLE
Bencoding

### PR DESCRIPTION
**Pull request by W1ndfan**
Fix info hash and switched to a new bencoding library

### CODE REVIEW

File: owltorrent-core/src/main/java/edu/rice/owltorrent/core/serialization/TorrentParser.java [hunk 1]
```diff
@@ -26,14 +27,16 @@
...
 
+  private static byte[] infoPart;
   /**
    * Bencode a Torrent file and extract important attributes
    *
    * @param file Torrent file path
    * @return pruned Torrent object
    */
-  public static Torrent parse(@NonNull File file) throws IOException {
+  public static Torrent parse(@NonNull File file) throws Exception {
```
Comment (maxyu1115) [inline owltorrent-core/src/main/java/edu/rice/owltorrent/core/serialization/TorrentParser.java:39]: why was this changed to generic Exception? (Note that we don't need to specify throwing RuntimeExceptions like BTException. )

File: owltorrent-common/src/test/java/edu/rice/owltorrent/common/util/EncryptorTest.java [hunk 2]
```diff
@@ -24,10 +24,15 @@ public void testSHA1EncryptString() {
...
+    byte[] bytes = new byte[] {(byte) 0xec};
+    byte[] c = new byte[head.length + bytes.length + 1];
+    System.arraycopy(head, 0, c, 0, head.length);
+    System.arraycopy(bytes, 0, c, head.length, bytes.length);
+    System.arraycopy(new String("e").getBytes(), 0, c, head.length + bytes.length, 1);
+    byte[] result = SHA1Encryptor.encrypt(c);
     String actualHexHash = new BigInteger(1, result).toString(16);
-
-    assertEquals(expectedHexHash.toLowerCase(), actualHexHash.toLowerCase());
+    System.out.println(actualHexHash);
```
Comment (maxyu1115) [inline owltorrent-common/src/test/java/edu/rice/owltorrent/common/util/EncryptorTest.java:35]: with only printing this is no longer a unit test....

File: owltorrent-core/src/main/java/edu/rice/owltorrent/core/serialization/TorrentParser.java [hunk 3]
```diff
@@ -61,21 +64,37 @@ public static Torrent parse(@NonNull File file) throws IOException {
    */
   public static Torrent extractAttributes(@NonNull Map<String, Object> dict) {
     String announceURL = (String) dict.get(announceField);
+    Map<String, Object> infoDict = (Map<String, Object>) dict.get(infoField);
 
-    var infoDict = (Map<String, Object>) dict.get(infoField);
     String name = (String) infoDict.get(nameField);
-    long pieceLength = (long) infoDict.get(pieceLengthField);
+    Long pieceLength = (Long) infoDict.get(pieceLengthField);
```
Comment (maxyu1115) [inline owltorrent-core/src/main/java/edu/rice/owltorrent/core/serialization/TorrentParser.java:70]: Why was this changed to object Long? would it ever be null?

File: owltorrent-common/src/main/java/edu/rice/owltorrent/common/util/Bencoder.java [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+import java.util.function.BiFunction;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * author: shijie, yuchen
```
Comment (maxyu1115) [inline owltorrent-common/src/main/java/edu/rice/owltorrent/common/util/Bencoder.java:19]: nit: this should be the @author tag after the paragraph

File: owltorrent-common/src/main/java/edu/rice/owltorrent/common/util/Exceptions.java [hunk 5]
```diff
@@ -14,4 +16,11 @@ public IllegalByteOffsets(String message) {
   }
 
   public static class ParsingTorrentFileFailedException extends Exception {}
+
+  @Getter
+  public static class BTException extends RuntimeException {
```
Comment (maxyu1115) [inline owltorrent-common/src/main/java/edu/rice/owltorrent/common/util/Exceptions.java:21]: Should this extend Runtime Exception or Exception/IOException? Note that RuntimeExceptions are all "sneakily thrown" and don't require people to catch it.

File: owltorrent-core/src/main/java/edu/rice/owltorrent/core/serialization/TorrentParser.java [hunk 6]
```diff
@@ -122,4 +142,54 @@ public static Torrent extractAttributes(@NonNull Map<String, Object> dict) {
 
     return fileLengths;
   }
+
+  private static String byteUrlEncode(byte[] bs) {
```
Comment (bhavrish) [inline owltorrent-core/src/main/java/edu/rice/owltorrent/core/serialization/TorrentParser.java:146]: No longer need these methods

File: owltorrent-common/src/main/java/edu/rice/owltorrent/common/entity/TwentyByteId.java [hunk 7]
```diff
@@ -39,10 +40,10 @@ public int hashCode() {
 
   @Override
   public String toString() {
-    return new String(bytes);
+    return new BigInteger(1, bytes).toString(16);
   }
 
   public static TwentyByteId fromString(String str) {
-    return new TwentyByteId(str.getBytes(StandardCharsets.US_ASCII));
+    return new TwentyByteId(str.getBytes(StandardCharsets.UTF_8));
```
[the remark this answers is not in the document]
Comment (W1ndfan) [reply owltorrent-common/src/main/java/edu/rice/owltorrent/common/entity/TwentyByteId.java:47]: Yes, I think so as the protocol states that all strings are UTF-8.

**Comment by yuchengu** [comment]
Change to store the raw pieces hash instead of encoded string which might causes problems when comparing hashes
